### DISCUSSION TITLE
Question about FBXLoader/GLTFLoader

### DISCUSSION THREAD

**Discussion by sunyonglincanada** [heremaps/harp.gl#2066] [Q&A] [answered]
Hi there,

I was trying to load a large fbx file using harp.gl but it did not render anything. Not sure I implemented it correctly. Can anyone take a look? Here is the codepen - https://codepen.io/sunyonglincanada/pen/ZEpZOQK

Besides, I've got another question that does harp.gl support threejs GLTF loader to load 3D model?

Thanks in advance!!

**Comment by nzjony** [comment, marked as answer]
@sunyonglincanada great question.

I'll check out the codepen and see what the issue is.

Yes, we have some internal examples that use the GLTF loader to load a 3D model in harp... let me pull out some code.

```
import { GLTF, GLTFLoader } from "three/examples/jsm/loaders/GLTFLoader";

...

const loader = new GLTFLoader();

...

The following code goes inside a custom data source, and it would most likely go inside the: `getTile(tileKey: TileKey): Tile` method.

const tile = new Tile(this, tileKey); //Or whatever Tile implementation you have.

// Assuming you know where to put the model....
const geoPoint = new GeoCoordinates(latitude, longitude, altitude);
const tangentSpace = {
    position: new THREE.Vector3(),
    xAxis: new THREE.Vector3(),
    yAxis: new THREE.Vector3(),
    zAxis: new THREE.Vector3()
};
this.projection.localTangentSpace(geoPoint, tangentSpace);
const worldPos = tangentSpace.position;
const scaleFactor = this.projection.getScaleFactor(worldPos);
worldPos.sub(tile.center);
let data : ArrayBuffer = <loaded model>
loader.parse(
               data,
                ".",
                (gltf: GLTF) => {
                    for (const child of gltf.scene.children) {
                        const obj: TileObject = child;
                        obj.displacement = worldPos;
                        obj.scale.setScalar(scaleFactor);
                        obj.setRotationFromMatrix(
                            new THREE.Matrix4().makeBasis(
                                tangentSpace.xAxis,
                                tangentSpace.yAxis,
                                tangentSpace.zAxis
                            )
                        );
                        obj.renderOrder = this.m_renderOrder ?? 1000;
                        obj.receiveShadow = tile.mapView.shadowsEnabled;
                        obj.castShadow = tile.mapView.shadowsEnabled;
                        if ((obj as any).material !== undefined) {
                            const alphaTest = 0.66;
                            const map = (obj as any).material.map;
                            const material = new THREE.MeshPhongMaterial({
                                map,
                                shininess: 60,
                                alphaTest,
                                polygonOffset: true, // prevent z-fighting at ground level
                                polygonOffsetFactor: -1.0,
                                polygonOffsetUnits: -1.0
                            });
                            (obj as any).material = material;

                            //Needed for shadows... can remove if you don't have shadows
                            const customDepthMaterial = new THREE.MeshDepthMaterial({
                                depthPacking: THREE.RGBADepthPacking,
                                map,
                                alphaTest
                            });

                            (obj as any).customDepthMaterial = customDepthMaterial;
                        }
                        registerTileObject(tile, obj, GeometryKind.Building);
                        if (obj.type === "Mesh") {
                            const mesh = obj as THREE.Mesh;
                            mesh.geometry.computeBoundingBox();
                            this.maxGeometryHeight = Math.max(
                                this.maxGeometryHeight,
                                mesh.geometry.boundingBox!.max.z
                            );
                        }
                        tile.objects.push(obj);
                    }
                    this.requestUpdate();
                    resolve();
                },
                (error: ErrorEvent) => {
                    logger.error(`Error parsing GLTF model`);
                    resolve();
                }
            );
```

Sorry @sunyonglincanada, we don't have any open example doing this, but I will make a task that we do this.

**Reply by nzjony** [reply]
@sunyonglincanada , regarding the large file, I get a console error:

`Failed to load resource: The network connection was lost.`

I also see that it downloads very slowly, see image: 
<img width="449" alt="Screenshot 2021-01-21 at 09 35 22" src="https://user-images.githubusercontent.com/2533428/105324813-2fe62500-5bcc-11eb-8670-1efdf1183a86.png">

Can you please try downloading the file and accessing it locally? You just need to put it in the `resources` directory.

**Reply by sunyonglincanada** [reply]
> @sunyonglincanada great question.
> 
> I'll check out the codepen and see what the issue is.
> 
> Yes, we have some internal examples that use the GLTF loader to load a 3D model in harp... let me pull out some code.
> 
> ```
> import { GLTF, GLTFLoader } from "three/examples/jsm/loaders/GLTFLoader";
> 
> ...
> 
> const loader = new GLTFLoader();
> 
> ...
> 
> The following code goes inside a custom data source, and it would most likely go inside the: `getTile(tileKey: TileKey): Tile` method.
> 
> const tile = new Tile(this, tileKey); //Or whatever Tile implementation you have.
> 
> // Assuming you know where to put the model....
> const geoPoint = new GeoCoordinates(latitude, longitude, altitude);
> const tangentSpace = {
>     position: new THREE.Vector3(),
>     xAxis: new THREE.Vector3(),
>     yAxis: new THREE.Vector3(),
>     zAxis: new THREE.Vector3()
> };
> this.projection.localTangentSpace(geoPoint, tangentSpace);
> const worldPos = tangentSpace.position;
> const scaleFactor = this.projection.getScaleFactor(worldPos);
> worldPos.sub(tile.center);
> let data : ArrayBuffer = <loaded model>
> loader.parse(
>                data,
>                 ".",
>                 (gltf: GLTF) => {
>                     for (const child of gltf.scene.children) {
>                         const obj: TileObject = child;
>                         obj.displacement = worldPos;
>                         obj.scale.setScalar(scaleFactor);
>                         obj.setRotationFromMatrix(
>                             new THREE.Matrix4().makeBasis(
>                                 tangentSpace.xAxis,
>                                 tangentSpace.yAxis,
>                                 tangentSpace.zAxis
>                             )
>                         );
>                         obj.renderOrder = this.m_renderOrder ?? 1000;
>                         obj.receiveShadow = tile.mapView.shadowsEnabled;
>                         obj.castShadow = tile.mapView.shadowsEnabled;
>                         if ((obj as any).material !== undefined) {
>                             const alphaTest = 0.66;
>                             const map = (obj as any).material.map;
>                             const material = new THREE.MeshPhongMaterial({
>                                 map,
>                                 shininess: 60,
>                                 alphaTest,
>                                 polygonOffset: true, // prevent z-fighting at ground level
>                                 polygonOffsetFactor: -1.0,
>                                 polygonOffsetUnits: -1.0
>                             });
>                             (obj as any).material = material;
> 
>                             //Needed for shadows... can remove if you don't have shadows
>                             const customDepthMaterial = new THREE.MeshDepthMaterial({
>                                 depthPacking: THREE.RGBADepthPacking,
>                                 map,
>                                 alphaTest
>                             });
> 
>                             (obj as any).customDepthMaterial = customDepthMaterial;
>                         }
>                         registerTileObject(tile, obj, GeometryKind.Building);
>                         if (obj.type === "Mesh") {
>                             const mesh = obj as THREE.Mesh;
>                             mesh.geometry.computeBoundingBox();
>                             this.maxGeometryHeight = Math.max(
>                                 this.maxGeometryHeight,
>                                 mesh.geometry.boundingBox!.max.z
>                             );
>                         }
>                         tile.objects.push(obj);
>                     }
>                     this.requestUpdate();
>                     resolve();
>                 },
>                 (error: ErrorEvent) => {
>                     logger.error(`Error parsing GLTF model`);
>                     resolve();
>                 }
>             );
> ```
> 
> Sorry @sunyonglincanada, we don't have any open example doing this, but I will make a task that we do this.

I've tried to use your code to render the GLTF file but got some undefined errors. Not sure if I am using it correctly, can you please take a look as well? Here is the gist  - https://gist.github.com/sunyonglincanada/0d381944acfbb4ce6f296e8811913721 Thanks in advance!!

**Reply by nzjony** [reply]
Can you please send me a screenshot of the errors?

Also, a gist makes it hard to reproduce, can you please make a github repo so I can just download it and run it?

From looking at the code, you need to remove `longitude` for example and provide an actual value.

Also you need to change `let data : ArrayBuffer = <loaded model>` To the actual loaded FBX model from disk.

**Reply by sunyonglincanada** [reply]
> Can you please send me a screenshot of the errors?
> 
> Also, a gist makes it hard to reproduce, can you please make a github repo so I can just download it and run it?
> 
> From looking at the code, you need to remove `longitude` for example and provide an actual value.
> 
> Also you need to change `let data : ArrayBuffer = <loaded model>` To the actual loaded FBX model from disk.

Thanks so much for your updates. I've created a github repo ( https://github.com/sunyonglincanada/harp.gl-GLTF ),  also I've changed the longitude and loaded glb file for the GLTF loader use.

**Reply by nzjony** [reply]
@sunyonglincanada , I fixed it, I led you down the wrong track, I saw in three.js there are some simple ways of loading GLTF models, see: https://github.com/sunyonglincanada/harp.gl-GLTF/pull/1

**Reply by sunyonglincanada** [reply]
> @sunyonglincanada , I fixed it, I led you down the wrong track, I saw in three.js there are some simple ways of loading GLTF models, see: [sunyonglincanada/harp.gl-GLTF#1](https://github.com/sunyonglincanada/harp.gl-GLTF/pull/1)

Thanks so much @nzjony. let me check the PR  and try it on my end.

**Reply by sunyonglincanada** [reply]
Hey @nzjony, I've checked on my end and it works fine using your method of loading GLTF models. I really appreciated your help with these back and forth discussions.

**Reply by nzjony** [reply]
@sunyonglincanada , glad it worked out and thanks for being involved.

**Reply by nzjony** [reply]
@sunyonglincanada , I look forward to see what cool stuff you can make with harp.gl and please reach out again if you have any troubles.

**Comment by sunyonglincanada** [comment]
Thanks @nzjony for the updates. I will try on my end and let you know if there is any issue.

**Reply by sunyonglincanada** [reply]
Hey @nzjony  Sorry for the late reply. I've tried to move my own test fbx file to /resources folder and it did load faster. But the issue is it did not render anything but got an error from FBXLoader.js (  https://imgur.com/a/iDylg4G ). Not sure if it has the fbx file max size limit. It works all good using dancing.fbx ( https://imgur.com/a/oAdIxjW  ). Can you please take a look?

Besides, I really appreciate the internal codes for GLTF loader you shared with me. For the code, GLTF loads the data to render the model. I am wondering if there is any relative code/example to load GLTF/GLB file directly using GLTF Loader pretty much like the FBXLoader way?

**Reply by nzjony** [reply]
@sunyonglincanada , no worries.

Ok, strange.

Can you please try to load your model using this example: https://github.com/mrdoob/three.js/blob/master/examples/webgl_loader_fbx.html

We need to first check if you are having an issue with three.js or harp.gl

**Reply by nzjony** [reply]
I don't quite understand your question:

> Besides, I really appreciate the internal codes for GLTF loader you shared with me. For the code, GLTF loads the data to render the model. I am wondering if there is any relative code/example to load GLTF/GLB file directly using GLTF Loader pretty much like the FBXLoader way?

Do you mean you would like someway in the API to pass in a url / local file path and have harp.gl do this for you?

**Reply by sunyonglincanada** [reply]
> https://github.com/mrdoob/three.js/blob/master/examples/webgl_loader_fbx.html

Thanks @nzjony for the updates. Let me try it on my end.

**Reply by sunyonglincanada** [reply]
> I don't quite understand your question:
> 
> > Besides, I really appreciate the internal codes for GLTF loader you shared with me. For the code, GLTF loads the data to render the model. I am wondering if there is any relative code/example to load GLTF/GLB file directly using GLTF Loader pretty much like the FBXLoader way?
> 
> Do you mean you would like someway in the API to pass in a url / local file path and have harp.gl do this for you?

Yup, you are correct. 😁  Is that doable based on the up to date harp.gl version?

**Reply by nzjony** [reply]
@sunyonglincanada , not yet, we would need to create the API, because you are the first person to request this. Please open an issue with what you want and then we can look to implement it.

**Reply by sunyonglincanada** [reply]
> @sunyonglincanada , not yet, we would need to create the API, because you are the first person to request this. Please open an issue with what you want and then we can look to implement it.

Got it, I will open an issue for it. Much appreciated!
